### DISCUSSION TITLE
NextAuth Prisma Adapter Type Conflict with Extended User Object

### DISCUSSION THREAD

**Discussion by Prashant-S29** [prisma/prisma#25892] [Q&A] [answered]
### Bug description

# NextAuth Prisma Adapter Type Conflict with Extended User Object

## Description
When attempting to extend the `User` interface in the `next-auth.d.ts` file to include a custom `role` property, a TypeScript type error occurs in the NextAuth configuration with the Prisma adapter.

## Environment
- NextAuth.js: 5.0.0-beta.25
- Prisma: ^5.14.0
- Next.js: ^15.0.1

### Detailed Reproduction Steps
1. Clone the repo `https://github.com/Prashant-S29/ici-fest-skit`  and setup the project
2. Go to `types/next-auth.d.ts`
3. Add the `role` property to the `User` interface
4. Run `npm run dev` or `yarn dev`
5. Check the TypeScript compilation errors
6. Observe the type conflict with the Prisma adapter

## Error Message
```
next-auth.d.ts(17, 5): 'role' is declared here.
index.d.ts(389, 5): The expected type comes from property 'adapter' which is declared here on type 'NextAuthConfig'
```

## Expected Behavior
The custom `role` property should be seamlessly integrated with the Prisma adapter without type conflicts.

## Actual Behavior
TypeScript throws a type error when the `role` property is added to the `User` interface, forcing removal of the custom property to resolve the conflict.

## Possible Workarounds
1. Remove the custom `role` property
2. Modify Prisma schema to include the `role` field
3. Use a type assertion or type casting (not recommended)

## Proposed Solution
[Describe potential solutions or ask for community input]

## Additional Notes
- Confirm if the issue persists across different versions of NextAuth and Prisma
- Verify Prisma schema compatibility with extended user types

## Reproducible Repository
[If possible, provide a link to a minimal reproducible repository]

## Related Issues
[Link to any related GitHub issues or discussions]

### Prisma information

```prisma

enum Role {
    ADMIN
    COORDINATOR
    USER
}

model User {
    id            String    @id @default(uuid())
    name          String?
    email         String?   @unique
    emailVerified DateTime?
    image         String?
    role          Role      @default(USER)
    accounts      Account[]
    sessions      Session[]
}

```

```ts
export const authConfig: NextAuthConfig = {
  adapter: PrismaAdapter(db),
  secret: env.AUTH_SECRET,
  session: {
    strategy: "jwt",
  },
  pages: {
    signIn: "/login",
  },

  providers: [
    CredentialsProvider({
```

### Environment & setup

- OS: Windows
- Database: PostgreSQL
- Node.js version: 22.7.0

### Prisma Version

```
prisma                  : 5.14.0
@prisma/client          : 5.14.0
```

**Comment by ludralph** [comment]
Hi @Prashant-S29 

Thank you for raising this issue. I'm unable to reproduce this issue. No typescript compilation errors were thrown. Can you please recheck your project and send very specific details to reproduce this?

**Comment by Prashant-S29** [comment]
Hey, here's the error ss

![image](https://github.com/user-attachments/assets/6f599fcb-58e2-4cee-8cf6-00581fb15091)

So the issue is, if i change the Credential back to default email and password, there's no error. check this

![image](https://github.com/user-attachments/assets/74508bbe-fa0b-4eaa-804d-552fca082c53)

But as soon as I change it according to my requirement, I see this error, check this

![image](https://github.com/user-attachments/assets/51685537-386b-4d69-8d0b-bc71d7e9a22e)

**Comment by hakancodes0** [comment]
// types/next-auth.d.ts
```ts
import "next-auth";
import "next-auth/jwt";

declare module "next-auth" {
  interface User {
    id: string;
    role: string;
  }
  interface Session {
    user: User;
  }
}

declare module "next-auth/jwt" {
  interface JWT {
    id: string;
    role: string;
  }
}
```

Is this your next-auth.d.ts file?

**Comment by Prashant-S29** [comment]
yes !

**Comment by ludralph** [comment, marked as answer]
Hi @Prashant-S29 

The authorize method in `CredentialsProvider` expects the `credentials` parameter to be of type `Partial<Record<...>>` by default, while your implementation expects a `Credential` object. This type mismatch leads to the error you're seeing.
In your authorize method:
```ts
async authorize(credentials: Credential | null)
```
You're using the `Credential` type, which expects properties like `adminId`, `password`, and `eventId` to be either `string` or `undefined`. The `credentials` object passed to this method is of type `Partial<Record<...>>`, where the values are `unknown` by default.

To fix this, you can adjust the `credentials` parameter within the authorize method to conform to the `Credential` type. TypeScript won't infer this automatically, so you must explicitly assert or validate it.
```ts
 async authorize(credentials) {

        if (!credentials) return null;

        const { adminId, eventId, password } = credentials as Credential; 
  ```
The line `const { adminId, eventId, password } = credentials as Credential`; explicitly tells TypeScript to treat credentials as a `Credential` object.

**Reply by Prashant-S29** [reply]
Hey @ludralph, thanks for the solution. I got it fixed by making following changes



### Before Code
```typescript
CredentialsProvider({
  type: "credentials",
  credentials: {
    adminId: { label: "Admin ID", type: "text" },
    password: { label: "Password", type: "password" },
    eventId: { label: "Event ID", type: "text" },
  },
  async authorize(credentials: Credential | null) {
    const { adminId, eventId, password } = credentials ?? {};
    // Handle admin login logic
  },
});
```

### After Code

```typescript
export default CredentialsProvider({
  type: "credentials",
  credentials: {
    adminId: { label: "Admin ID", type: "text" },
    password: { label: "Password", type: "password" },
    eventId: { label: "Event ID", type: "text" },
  },
  async authorize(
    // updated type
    credentials: Partial<Record<keyof Credential, unknown>> | null,
  ) {
    if (!credentials) return null;

    // Type assertion
    const adminId = credentials.adminId as string | undefined;
    const password = credentials.password as string | undefined;
    const eventId = credentials.eventId as string | undefined;
```
